src/identity.js

```diff
@@ -511,16 +537,36 @@ export class Identity extends EventEmitter {
                 throw err;
             }
 
-            if (sessionData && this._enableSessionCaching) {
-                const expiresIn = 1000 * (sessionData.expiresIn || 300);
-                this.cache.set(HAS_SESSION_CACHE_KEY, sessionData, expiresIn);
+            if (sessionData){
+                // for expiring session and safari browser do full page redirect to gain new session
+                if(_checkRedirectionNeed(sessionData)){
+                    if (this._enableSessionCaching) {
```

Comment:
Can this "semaphore" logic be extracted into separate methods?
`hasSession` does a lot already.